File: 15/Solution.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class Solution
{
    static class Pair
    {
        int first, second;

        public Pair(int first, int second)
        {
            this.first = first;
            this.second = second;
        }
    }

    static class Node
    {
        int x, y, s;

        public Node(int x, int y, int s)
        {
            this.x = x;
            this.y = y;
            this.s = s;
        }
    }

    public static final int[] dx = { 1, -1, 0, 0 };
    public static final int[] dy = { 0, 0, 1, -1 };

    public static void main(String[] args)
    {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();

        int[][] mz = new int[1005][1005];
        Pair[][] pw = new Pair[1005][1005];

        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                mz[i][j] = sc.nextInt();
                pw[i][j] = new Pair(0, -1);
            }
        }

        Queue<Node> q = new LinkedList<>();
        pw[0][0] = new Pair(1, 0);
        q.add(new Node(0, 0, 0));

        while (!q.isEmpty())
        {
            Node k = q.poll();
            for (int i = 0; i < 4; i++)
            {
                int nx = dx[i] + k.x;
                int ny = dy[i] + k.y;
                if (nx >= 0 && nx < m && ny >= 0 && ny < n && mz[nx][ny] != 1)
                {
                    if (pw[nx][ny].second == -1)
                    {
                        pw[nx][ny] = new Pair(pw[k.x][k.y].first, k.s);
                        q.add(new Node(nx, ny, k.s + 1));
                    }
                    else if (k.s == pw[nx][ny].second)
                    {
                        pw[nx][ny].first += pw[k.x][k.y].first;
                        pw[nx][ny].first %= 1000000007;
                    }
                }
            }
        }
        System.out.println(pw[m - 1][n - 1].first);
    }
}

File: 15/solution.cpp
#include <bits/stdc++.h>
#define umap unordered_map
using namespace std;
int n, m;
umap<int, umap<long long, pair<int, int>>> pw;
short int mz[1005][1005];
struct node
{
    int x, y;
    int s;
};
int dx[4] = {1, -1, 0, 0};
int dy[4] = {0, 0, 1, -1};
int main()
{
    cin >> m >> n;
    for (int i = 0; i < m; i++)
    {
        for (int j = 0; j < n; j++)
        {
            cin >> mz[i][j];
            pw[i][j] = {0, -1};
        }
    }
    queue<node> q;
    pw[0][0] = {1, 0};
    q.push({0, 0, 0});
    while (!q.empty())
    {
        node k = q.front();
        for (int i = 0; i < 4; i++)
        {
            int nx = dx[i] + k.x;
            int ny = dy[i] + k.y;
            if (nx >= 0 && nx < m && ny >= 0 && ny < n && mz[nx][ny] != 1)
            {
                if (pw[nx][ny].second == -1)
                {
                    pw[nx][ny] = {pw[k.x][k.y].first, k.s};
                    q.push({nx, ny, k.s + 1});
                }
                else if (k.s == pw[nx][ny].second)
                {
                    pw[nx][ny].first += pw[k.x][k.y].first;
                    pw[nx][ny].first %= 1000000007;
                }
            }
        }
        q.pop();
    }
    cout << pw[m - 1][n - 1].first << endl;
    return 0;
}